`EditorLauncher` Scaffold Implementation

oguzkocer opened the pull request:
## Summary
Adds a centralized `EditorLauncher` system to replace direct Intent creation for editor activities, providing type-safe parameters and consistent analytics tracking.

## Motivation
This scaffolding enables the future introduction of a dedicated `GutenbergKitActivity` by centralizing editor launch logic. The plan is to:

1. **Duplicate EditPostActivity** as `GutenbergKitActivity` 
2. **Simplify GutenbergKitActivity** to only work with GutenbergKit editor, removing legacy code paths
3. **Use EditorLauncher routing** to direct users to the appropriate activity based on feature flags
4. **Keep EditPostActivity** as the legacy activity for existing functionality

This approach allows iterating on a clean, modern editor experience while maintaining backward compatibility through the existing activity.

## Key Components

### EditorLauncher
- **Purpose**: Centralized helper for launching editor activities with proper routing logic
- **Features**:
  - Determines which editor activity to launch based on feature flags
  - Provides analytics tracking for editor launches
  - Handles all Intent extra mapping from type-safe parameters
- **Analytics**: Tracks `EDITOR_LAUNCHER` events with `will_use_gutenberg_kit` property and adds `EXTRA_LAUNCHED_VIA_EDITOR_LAUNCHER` intent extra

### EditorLauncherParams
- **Purpose**: Type-safe parameters for launching editor activities
- **Design**: Kotlin data class with Java Builder pattern for compatibility
- **Builder**: Requires non-null SiteModel in constructor for type safety
- **Coverage**: Supports all existing editor launch scenarios (posts, pages, reblog, etc.)

### Integration
- **ActivityLauncher.java**: Updated all editor launch methods to use EditorLauncher
- **Analytics**: Centralized `EXTRA_CREATION_SOURCE_DETAIL` handling eliminates code duplication
- **Dependency Injection**: EditorLauncher registered in AppComponent with singleton scope

## Technical Details

### Type Safety Improvements
- Builder pattern requires `SiteModel` parameter upfront (no runtime null checks)
- All editor parameters validated at compile time through structured data class
- Intent extras automatically mapped from parameters

### Analytics Centralization
Previously, every ActivityLauncher method manually added analytics extras:
```java
Intent intent = EditorLauncher.getInstance().createEditorIntent(activity, params);
intent.putExtra(AnalyticsUtils.EXTRA_CREATION_SOURCE_DETAIL, source);
```

Now handled automatically in EditorLauncher when `params.source` is provided.

### Launch Method Tracking
EditorLauncher adds `EXTRA_LAUNCHED_VIA_EDITOR_LAUNCHER` intent extra to distinguish between:
- Editor launches via EditorLauncher (tracked as `EDITOR_LAUNCHED_VIA_EDITOR_LAUNCHER`)  
- Direct Intent creation (legacy path)

This enables measuring EditorLauncher adoption vs legacy launch methods.

### Completeness Validation
- **EditorLauncherTest**: Unit test validates all EditorLauncherParams fields are handled
- **Documentation**: Cross-references between classes and test for field-to-method mapping
- **Static Enforcement**: Test fails if new fields aren't properly handled

## Backward Compatibility
- All existing editor launch paths continue to work unchanged
- Legacy direct Intent creation still supported where needed
- No changes to editor activity implementations

oguzkocer committed:
Add EditorLauncher scaffold with centralized editor routing

Implement hybrid approach for editor activity routing that will support
both legacy EditPostActivity and future EditPostGutenbergKitActivity.

Changes:
- Add EditorLauncher singleton with feature-flag-based routing logic
- Migrate all ActivityLauncher methods to use EditorLauncher helper
- Add EditorLauncher to dependency injection (AppComponent)
- Include analytics tracking and source identification for launches
- Route to EditPostActivity for now (scaffold implementation)

This provides infrastructure for clean editor activity separation while
maintaining backward compatibility. Future changes will:
- Add EditorLauncherParams data class with builder pattern
- Route to EditPostGutenbergKitActivity when implemented
- Add redirection fallback with monitoring

Covers ActivityLauncher methods:
- openEditorForSiteInNewStack()
- openEditorForPostInNewStack()
- openEditorForReblog()
- addNewPostForResult() variants
- editPostOrPageForResult() variants
- editPageForResult() variants
- addNewPageForResult() variants

oguzkocer committed:
Add EditorLauncherParams for type-safe editor launch parameters

Replace Bundle-based Intent extras with type-safe parameter objects
to improve maintainability and reduce errors.

Changes:
- Add EditorLauncherParams.kt with Kotlin data class and Java Builder pattern
- Update EditorLauncher.kt to accept EditorLauncherParams instead of Intent bundles
- Refactor ActivityLauncher.java methods to use EditorLauncherParams.Builder
- Add backward compatibility with deprecated Intent-based method
- Centralize all Intent extra handling in EditorLauncher.addEditorExtras()
- Maintain analytics tracking with type-safe parameter source detection

This provides compile-time safety while maintaining Java interoperability
through the builder pattern.

oguzkocer committed:
Remove deprecated Intent-based EditorLauncher method

Complete migration to EditorLauncherParams for all ActivityLauncher methods.

Changes:
- Remove deprecated createEditorIntent(context, Intent) method from EditorLauncher
- Remove unused getIntentSource() legacy method
- Update remaining ActivityLauncher methods to use EditorLauncherParams.Builder:
  - addNewPostWithContentFromAIForResult()
  - addNewPageForResult() (both Activity and Fragment variants)
- All editor launches now use type-safe parameters instead of Bundle operations

This completes the EditorLauncherParams migration for ActivityLauncher,
providing compile-time safety and better maintainability.

oguzkocer committed:
Fix Detekt code quality violations in EditorLauncher

Address all code quality issues identified by Detekt static analysis.

Changes:
- Remove unused imports (Bundle) and unused property (analyticsTracker)
- Replace TODO comments with regular comments to avoid ForbiddenComment violations
- Break down complex addEditorExtras() method into smaller helper methods:
  - addBasicExtras() - site, page, promo flags
  - addPostExtras() - post IDs, autosave, quickpress, landing editor flags
  - addReblogExtras() - reblog title, quote, image, citation, action
  - addPageExtras() - page title, content, template
  - addMiscExtras() - voice content, media, prompt ID, entry point
- Fix MaxLineLength violations by splitting long builder method calls
- Add missing newline at end of EditorLauncherParams.kt

This reduces cyclomatic complexity and improves code maintainability
while maintaining all existing functionality.

oguzkocer committed:
Move analytics source field handling into EditorLauncher

Centralizes the EXTRA_CREATION_SOURCE_DETAIL Intent extra creation in
EditorLauncher.addMiscExtras instead of requiring manual addition in
each ActivityLauncher method. This eliminates code duplication and
ensures analytics tracking is consistently applied across all editor launches.

Changes:
- Add AnalyticsUtils.EXTRA_CREATION_SOURCE_DETAIL to EditorLauncher.addMiscExtras
- Remove redundant analytics field additions from ActivityLauncher methods
- Add AnalyticsUtils import to EditorLauncher

oguzkocer committed:
Require non-null SiteModel in EditorLauncherParams.Builder constructor

Makes the Java Builder pattern more type-safe by requiring SiteModel in the constructor
rather than checking for null at build time.

Changes:
- EditorLauncherParams.Builder now takes @NonNull SiteModel in constructor
- Remove site() setter method from Builder
- Remove runtime requireNotNull check in build()
- Update all call sites to pass site to constructor
- Add missing @NonNull annotations in ActivityLauncher

oguzkocer committed:
Add analytics tracking to EditorLauncher

Tracks editor launches with should_use_gutenberg_kit property to understand
feature flag usage patterns.

oguzkocer committed:
Add documentation and test for EditorLauncherParams field mapping

Ensures all EditorLauncherParams fields are handled in EditorLauncher.addEditorExtras()
through documentation and a unit test that validates completeness.

Changes:
- Add cross-reference documentation in EditorLauncherParams and EditorLauncher
- Add EditorLauncherTest with detailed field-to-method mapping documentation
- Test validates that all fields are accounted for in addEditorExtras methods
- Filter out Kotlin synthetic fields in test validation

oguzkocer committed:
Fix line length in EditorLauncherTest

oguzkocer committed:
Update analytics property name to will_use_gutenberg_kit

oguzkocer committed:
Add tracking for EditorLauncher vs direct Intent launches

Adds a way to distinguish editor launches that went through EditorLauncher
vs direct Intent creation for analytics purposes.

Changes:
- EditorLauncher: Add EXTRA_LAUNCHED_VIA_EDITOR_LAUNCHER intent extra
- EditPostActivity: Track EDITOR_LAUNCHED_VIA_EDITOR_LAUNCHER when extra is present (only on initial creation, not configuration changes)
- AnalyticsTracker: Add EDITOR_LAUNCHED_VIA_EDITOR_LAUNCHER stat

oguzkocer committed:
Fix line length in EditPostActivity

nbradbury commented:
I asked Claude this:

> Is EditPostActivity ever launched from somewhere other than EditorLauncher?

And got this response:

```
Yes, EditPostActivity is still launched directly from 4 other places that weren't migrated 
to use EditorLauncher:

  1. ShareIntentReceiverFragment.java:214 - Enum definition for share actions
  2. AddQuickPressShortcutActivity.java:143 - Creating QuickPress shortcuts
  3. UploadUtils.java:613 - Launching editor after media upload
  4. PostUploadNotifier.java:444 - Notification action to edit post

  These should also be migrated to use EditorLauncher for consistency 
```

I don't think the first one is an issue, but the others appear to be launching `EditPostActivity`. Should this be addressed in this PR?

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java:
```diff
@@ -1023,14 +1041,25 @@
 
-    public static void editPostOrPageForResult(Activity activity, SiteModel site, PostModel post) {
-        editPostOrPageForResult(new Intent(activity, EditPostActivity.class), activity, site, post.getId(), false);
+    public static void editPostOrPageForResult(Activity activity, @NonNull SiteModel site, PostModel post) {
```
Is there any concern that the `PostModel` could be null? According to Claude:

```
 Yes, editPostOrPageForResult can be called with a null post.

  Looking at the evidence:

  1. WPMainActivity.java:1117-1118:
  PostModel post = mPostStore.getPostByLocalPostId(postId);
  ActivityLauncher.editPostOrPageForResult(this, selectedSite, post);
    - getPostByLocalPostId returns null if no post is found (line 601-602 in PostStore.java)
    - The post is passed directly without null checking
  2. PostListAction.kt: The EditPost action class requires a non-null PostModel in its constructor, so this path is safe
  3. UploadUtils.java: The method signature has @NonNull final PostModel post, so callers must ensure non-null

  The issue: The new implementation in the PR calls post.getId() directly without null checking:
  public static void editPostOrPageForResult(Activity activity, @NonNull SiteModel site, PostModel post) {
      EditorLauncherParams params = new EditorLauncherParams.Builder(site)
              .postLocalId(post.getId())  // <- NullPointerException if post is null
              .loadAutoSaveRevision(false)
              .build();

  This is a bug in the PR. The method should either:
  1. Add @NonNull annotation to the post parameter and let callers handle null
  2. Add null checking like:
  if (post == null) {
      return;
  }

  The safest fix would be to add null checking since at least one caller (WPMainActivity) doesn't check for null before calling.
```

oguzkocer replied:
I think this issue is present in `trunk`, but I didn't have a chance to fully investigate it. Regardless, it seems like a legitimate concern, so I added a check for this: 29c3d84bd50027b981e9e68731875f15f54fdb80

nbradbury commented:
Sharing a photo to the app also fails with the same `HTTP 400` issue.

```
Unexpected response code 400 for https://public-api.wordpress.com/rest/v1.1/sites/235002728/post/0/diffs/?locale=en_US
```

https://github.com/user-attachments/assets/ac9053a0-f06c-405b-ae97-fd37928df49f

dcalhoun reviewed (commented):
> When I tap a post in the post list, it shows an empty post in the editor. Same for pages.

I experience the same for each of the three editors—GutenbergKit, Gutenberg Mobile, and Aztec. Surprisingly, it does not occur when creating a new page from a template; that flow successfully populates the editor. 

> Sharing a photo to the app also fails with the same `HTTP 400` issue.

Sharing media to GutenbergKit is currently unsupported, see CMM-190. That said, sharing media to Gutenberg Mobile appears broken in the `trunk` branch, so likely unrelated to these changes. 😞  

---

I performed quick smoke testing of fundamental features in each of GutenbergKit, Gutenberg Mobile, and Aztec—text editing, media uploads, drafting, publishing, analytics, network connection changes, OS appearance changes, etc. I did not encounter regressions aside from those noted above.

oguzkocer committed:
Add null checks for PostModel in editPostOrPageForResult methods

Prevents crashes when PostModel is null (e.g., when mPostStore.getPostByLocalPostId
returns null for non-existent posts) by showing a user-friendly error message instead.

Changes:
- Add null check and error toast in both editPostOrPageForResult variants
- Prevents crashes from calling post.getId() or post.isPage() on null objects

oguzkocer committed:
Make all boolean parameters nullable to avoid unintended defaults

The trunk editPostOrPageForResult methods never set boolean extras like
EXTRA_IS_PAGE, EXTRA_IS_PROMO, or EXTRA_LOAD_AUTO_SAVE_REVISION unless
explicitly required. Making all boolean parameters nullable ensures they're
only set when explicitly provided, matching trunk behavior exactly.

Changes:
- Make all boolean fields nullable in EditorLauncherParams and Builder
- Update EditorLauncher to only set extras when parameters are non-null
- Remove unnecessary loadAutoSaveRevision(false) from first editPostOrPageForResult
- Preserves explicit boolean settings in other methods (pages, promos, etc.)

oguzkocer commented:
> I don't think the first one is an issue, but the others appear to be launching EditPostActivity. Should this be addressed in this PR?

@nbradbury Let me take a look at that in a separate PR.

oguzkocer committed:
Don't set source in ActivityLauncher.openEditorForReblog

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java:
```diff
@@ -456,12 +461,16 @@ public static void openEditorForReblog(
+                .reblogPostImage(post.getFeaturedImage())
+                .reblogPostCitation(post.getUrl())
+                .reblogAction(EditPostActivityConstants.ACTION_REBLOG)
+                .source(reblogSource)
```
The original implementation didn't set this here and set it in `addNewPostForResult` instead. This wouldn't have been a problem, but I've removed it in 4198b77427c11c0b9e4d7d3de08d02f5621d2770 to match the previous implementation.

nbradbury reviewed (approved):
@oguzkocer The problem with opening posts/pages appears to be resolved so merge when ready :shipit:

dcalhoun reviewed (approved):
The latest iteration tested well for me in GutenbergKit, Gutenberg Mobile, and Aztec. I believe we are good to merge.
